scripts/ci/dag.py

```diff
@@ -0,0 +1,169 @@
+        return self._num_finished == len(self._nodes)
+
+    def walk(self, f: Callable[[_T], None], max_tokens: int, refill_interval_s: float) -> None:
+        """
+        Process the graph in parallel.
+
+        Each node in the graph is processed only once all of its dependencies have been processed.
+
+        Concurrency is limited by the following bucket rate limiting algorithm:
+        * Processing may not begin until a token can be acquired from the bucket.
+        * There are at most `max_tokens - in_progress_tasks` in the bucket at any time.
+        * Tokens are refreshed every `refill_interval_s`.
+        """
+
+        num_cpus = cpu_count()
+        # we need one main thread + N worker threads, so `N = num_cpus - 1`,
+        # and we need at least 1 worker thread.
+        num_workers = num_cpus - 1 if num_cpus > 0 else 1
+        with ThreadPoolExecutor(max_workers=num_workers) as p:
+            task_queue: Queue[_T] = Queue()
+            done_queue: Queue[_T] = Queue()
+            shutdown: EventClass = Event()
+
+            def worker(n: int) -> None:
+                # Attempt to grab a task from the queue,
+                # execute it, then put it in the done queue.
+                while not shutdown.is_set():
+                    try:
+                        node = task_queue.get_nowait()
+                        f(node)
+                        done_queue.put(node)
+                    except Empty:
+                        time.sleep(0)  # yield to prevent busy-looping
+                        continue
+
+            for n in range(0, num_workers):  # start all workers
+                p.submit(worker, n)
+
+            tokens = max_tokens
+            in_progress = 0
+            last_refill = time.time()
+            while not self._is_done():
+                # This loop has two parts, `push` and `pull`.
+                #
+                # The `push` loop attempts to push tasks
+                # onto the `task_queue` while there are
+                # some tasks ready to go, and some tokens left.
+                # It is also responsible for refreshing the bucket.
+                #
+                # The `pull` loop attempts to retrieve done
+                # tasks and decrement the dependency counter on
+                # their dependents.
+                #
+                # Once a node has no pending dependencies left,
+                # it becomes ready and will be queued in one of
+                # the iterations of the `push` loop.
+                #
+                # It's important to always use non-blocking `get`
+                # with the task and done queues.
+                # This is so that both the push and pull loops
+                # can eventually make progress.
+                while len(self._queue) > 0:  # push loop
+                    now = time.time()
+                    if now - last_refill > refill_interval_s:
```

Comment:
things are often easier to read without negations

```suggestion
                    if now > last_refill + refill_interval_s:
```

and easier still when ordered as numbers on a number line


```suggestion
                    if last_refill + refill_interval_s < now:
```

Reply:
For me, there isn't a big difference in readability between the three options. I default to the first one because I tend to think about it in terms of delta time rather than time instants.

Reply:
ah, well, ymmv :)

Reply:
at least we don't get underflow here with negation as we could easily do in Rust 😬